mdtk/tests/test_midi.py

```diff
@@ -0,0 +1,199 @@
+import pandas as pd
+import os
+import pretty_midi
+import shutil
+
+import mdtk.midi as midi
+
+USER_HOME = os.path.expanduser('~')
+TEST_CACHE_PATH = os.path.join(USER_HOME, '.mdtk_test_cache')
+MIDI_PATH = f"mdtk{os.path.sep}tests{os.path.sep}test.mid"
+
+def test_df_to_csv():
+    notes = []
+    notes.append({'onset': 0.0,
+                  'track': 1,
+                  'pitch': 42,
+                  'dur': 1.5})
+    notes.append({'onset': 0.3,
+                  'track': 1,
+                  'pitch': 40,
+                  'dur': 1.07})
+    notes.append({'onset': 0.3,
+                  'track': 2,
+                  'pitch': 56,
+                  'dur': 1.11})
+    
+    df = pd.DataFrame(notes)
+    
+    csv_name = TEST_CACHE_PATH + os.path.sep + 'test.csv'
+    try:
+        os.remove(csv_name)
+    except:
+        pass
+    
+    midi.df_to_csv(df, csv_name)
+    
+    assert os.path.exists(csv_name), ('No csv created.')
+    
+    # Check that notes were written correctly
+    with open(csv_name, 'r') as file:
+        for i, line in enumerate(file):
+            split = line.split(',')
+            assert float(split[0]) == notes[i]['onset'], ("Onset time of " +
+                   f"note {i} ({notes[i]}) not equal to csv's written onset " +
+                   f"time of {float(split[0])}")
+            assert split[1].isdigit(), f"Track {split[1]} is not an int."
+            assert int(split[1]) == notes[i]['track'], ("Track of " +
+                   f"note {i} ({notes[i]}) not equal to csv's written track " +
+                   f"of {int(split[1])}")
+            assert split[2].isdigit(), f"Pitch {split[2]} is not an int."
+            assert int(split[2]) == notes[i]['pitch'], ("Pitch of " +
+                   f"note {i} ({notes[i]}) not equal to csv's written pitch " +
+                   f"of {int(split[2])}")
+            assert float(split[3]) == notes[i]['dur'], ("Duration of " +
+                   f"note {i} ({notes[i]}) not equal to csv's written " +
+                   f"duration of {float(split[3])}")
+            
+            
+            
+def test_midi_to_df():
+    df = midi.midi_to_df(MIDI_PATH)
+    
+    # This relies on pretty_midi being correct
+    m = pretty_midi.PrettyMIDI(MIDI_PATH)
+    midi_notes = []
+    for i, instrument in enumerate(m.instruments):
+        for note in instrument.notes:
+            midi_notes.append({'onset': note.start * 1000,
+                               'track': i,
+                               'pitch': note.pitch,
+                               'dur': (note.end - note.start) * 1000})
+    
+    df_notes = []
+    for index, df_note in df.iterrows():
+        # There must be a better way to do this
```

Comment:
I think `df.to_dict('records')` does what you want: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_dict.html

Reply:
Yup, that's the one!